File: quant_rl_trading/dashboard/services/agent_health.py
"""Agent Health 집계 — Analyst 가 살아 있는지, 그리고 쓸모가 있는지.

Data Quality 가 "데이터가 썩고 있나" 를 본다면 이 화면은 **"에이전트가 썩고
있나"** 를 본다. 둘은 다른 질문이다. 데이터가 완벽해도 Analyst 의 IC 는
시간이 지나면 감쇠한다(알파 소멸). 그걸 못 보면 죽은 신호에 계속 가중치를
주게 된다.

여기서 가장 중요한 숫자는 IC 가 아니라 **가중치**다. IC 0.03 을 넘지 못한
Analyst 는 가중치 0(관찰 모드)으로 동작한다 — 점수는 계속 기록하되 매매에는
쓰지 않는다. 그 상태가 화면에 상시 보여야, 검증되지 않은 것이 조용히 켜져
있는 일이 생기지 않는다.

**없는 것은 없다고 말한다.** 아직 측정하지 않은 Analyst, 아직 비어 있는
Verdict 성적표는 0 으로 채우지 않고 비어 있음을 그대로 돌려준다. 0 과
'모름' 은 다른 사실이다.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

import pandas as pd

from quant_rl_trading.analysts import scorecard
from quant_rl_trading.store import Store

WEIGHTS = "analyst_weights"
SIGNALS = "signals"
VERDICTS = "verdicts"

#: 아직 구현되지 않았거나 데이터가 없어 관찰 모드인 Analyst.
#: 명단에서 빼지 않는다 — 빠지면 "왜 없지" 를 아무도 묻지 않게 된다.
PLANNED = {
    "chart": "가격 추세",
    "flow_kr": "투자자별 수급 (LS t1717)",
    "flow_us": "미장 수급 — 데이터 없음",
    "fundamental": "DART 재무",
    "news": "공시·뉴스 필터 (Verdict)",
    "sns": "펌핑 탐지 (Verdict)",
    "regime": "지수·변동성",
    "event": "달력",
    "risk": "상관·변동성·유동성",
    "volume": "거래량 급증 (chart 에서 분리)",
}


def latest_weights(store: Store, *, as_of: datetime, lookback: int) -> pd.DataFrame:
    """Analyst 별 가장 최근 측정 결과.

    같은 Analyst 를 여러 번 측정하면 행이 쌓인다 (append-only). 현재 상태는
    그중 가장 늦은 ``valid_from`` 이다.
    """
    frame = store.get(WEIGHTS, as_of=as_of, lookback=lookback)
    if frame.empty:
        return frame
    return frame.sort_values("valid_from").groupby(["entity_id", "market"]).tail(1)


def roster(store: Store, *, as_of: datetime, lookback: int) -> list[dict[str, Any]]:
    """Analyst 전원의 상태. 측정 안 된 것도 명단에 남긴다."""
    measured = latest_weights(store, as_of=as_of, lookback=lookback)
    by_name: dict[str, dict[str, Any]] = {}
    if not measured.empty:
        for row in measured.to_dict(orient="records"):
            by_name[str(row["entity_id"])] = row

    out: list[dict[str, Any]] = []
    for name, note in sorted(PLANNED.items()):
        row = by_name.get(name)
        if row is None:
            out.append(
                {
                    "analyst": name,
                    "note": note,
                    "measured": False,
                    # 측정 전에는 가중치가 0 이다. "아직 모름" 이 아니라
                    # "아직 자격 없음" 이 맞다 — 검증 전에는 매매에 쓰지 않는다.
                    "weight": 0.0,
                    "ic": None,
                    "passed": None,
                    "sample_days": None,
                    "version": None,
                    "market": None,
                    "measured_at": None,
                }
            )
            continue
        out.append(
            {
                "analyst": name,
                "note": note,
                "measured": True,
                "weight": float(row["weight"]),
                "ic": float(row["ic"]),
                "ic_threshold": float(row["ic_threshold"]),
                "passed": bool(row["passed"]),
                "sample_days": int(row["sample_days"]),
                "version": str(row["analyst_version"]),
                "market": str(row["market"]),
                "measured_at": row["valid_from"].isoformat(),
            }
        )
    return out


def ic_history(store: Store, *, as_of: datetime, lookback: int) -> dict[str, Any]:
    """Analyst 별 IC 측정 이력.

    한 점만 있으면 추이가 아니다. 그래도 그리는 이유는 **감쇠를 보기 위해**서다
    — 재측정이 쌓이면 IC 가 내려가는 것이 여기서 먼저 보인다 (알파 소멸).
    """
    frame = store.get(WEIGHTS, as_of=as_of, lookback=lookback)
    if frame.empty:
        return {"series": [], "points": 0}

    series: list[dict[str, Any]] = []
    for name, group in frame.sort_values("valid_from").groupby("entity_id"):
        series.append(
            {
                "analyst": str(name),
                "points": [
                    {
                        "at": row["valid_from"].isoformat(),
                        "ic": float(row["ic"]),
                        "threshold": float(row["ic_threshold"]),
                        "passed": bool(row["passed"]),
                    }
                    for row in group.to_dict(orient="records")
                ],
            }
        )
    return {"series": series, "points": len(frame)}


def signal_activity(store: Store, *, as_of: datetime, lookback: int) -> dict[str, Any]:
    """Signal 기록 현황. '점수를 내고 있는가' 를 본다.

    가중치 0(관찰 모드)이어도 **Signal 은 계속 기록돼야 한다.** 기록이 멈추면
    나중에 그 Analyst 를 켤 근거를 만들 수 없다.
    """
    frame = store.get(SIGNALS, as_of=as_of, lookback=lookback)
    if frame.empty:
        return {"analysts": [], "total": 0}

    rows: list[dict[str, Any]] = []
    for (name, version), group in frame.groupby(["analyst", "analyst_version"]):
        latency = group["latency_ms"].astype(float)
        rows.append(
            {
                "analyst": str(name),
                "version": str(version),
                "signals": len(group),
                "entities": int(group["entity_id"].nunique()),
                "sessions": int(group["valid_from"].dt.date.nunique()),
                "last_as_of": group["valid_from"].max().isoformat(),
                "latency_p50_ms": float(latency.quantile(0.5)),
                "latency_p90_ms": float(latency.quantile(0.9)),
                "mean_confidence": float(group["confidence"].astype(float).mean()),
            }
        )
    return {"analysts": sorted(rows, key=lambda item: str(item["analyst"])), "total": len(frame)}


def verdict_scorecard(store: Store, *, as_of: datetime, lookback: int) -> dict[str, Any]:
    """News·SNS 거부 성적표.

    차단한 종목의 **이후 수익률**을 추적한다. 필터가 실제로 손실을 피하게
    해 줬는지 사후에 따지지 않으면, 이 필터는 검증되지 않은 채로 매수만
    막는 장치가 된다.

    거부는 IC 로 검증할 수 없다(점수가 아니라 판정이므로). 그래서 이 성적표가
    유일한 검증 수단이다.
    """
    frame = store.get(VERDICTS, as_of=as_of, lookback=lookback)
    if frame.empty:
        return {
            "blocks": 0, "by_category": [], "by_analyst": [], "active": 0,
            "scorecard": scorecard.evaluate_blocks(store, as_of=as_of, lookback=lookback),
        }

    blocked = frame[frame["decision"] == "block"]
    active = blocked[blocked["expires_at"] > as_of]

    return {
        "blocks": len(blocked),
        "active": len(active),
        # 차단이 실제로 손실을 피하게 해 줬는지. IC 를 못 쓰는 이 둘의
        # 유일한 검증 수단이다.
        "scorecard": scorecard.evaluate_blocks(store, as_of=as_of, lookback=lookback),
        "by_category": [
            {"category": str(name), "count": int(count)}
            for name, count in blocked["category"].value_counts().items()
        ],
        "by_analyst": [
            {"analyst": str(name), "count": int(count)}
            for name, count in blocked["analyst"].value_counts().items()
        ],
    }


def summary(
    store: Store, *, as_of: datetime, lookback: int, thresholds: dict[str, Any]
) -> dict[str, Any]:
    people = roster(store, as_of=as_of, lookback=lookback)
    activity = signal_activity(store, as_of=as_of, lookback=lookback)
    verdicts = verdict_scorecard(store, as_of=as_of, lookback=lookback)

    passed = [item for item in people if item["passed"]]
    measured = [item for item in people if item["measured"]]

    return {
        "total": len(people),
        "measured": len(measured),
        "passed": len(passed),
        "observing": len(people) - len(passed),
        "active_weight": sum(float(item["weight"]) for item in people),
        "signals": activity["total"],
        "blocks": verdicts["blocks"],
        "warnings": _warnings(people, activity, thresholds),
    }


def _warnings(
    people: list[dict[str, Any]],
    activity: dict[str, Any],
    thresholds: dict[str, Any],
) -> list[str]:
    found: list[str] = []
    passed = [item for item in people if item["passed"]]

    # M2 완료 기준: 최소 2개 Analyst 가 IC 0.03 통과.
    if len(passed) < 2:
        found.append(
            f"IC 통과 Analyst {len(passed)}명 — M2 완료 기준은 2명이다"
        )
    if not activity["analysts"]:
        found.append("기록된 Signal 이 없다. 관찰 모드여도 점수는 남겨야 한다")

    unmeasured = [item["analyst"] for item in people if not item["measured"]]
    if unmeasured:
        found.append(f"미측정 {len(unmeasured)}명: {', '.join(unmeasured)}")
    return found


__all__ = [
    "PLANNED",
    "ic_history",
    "latest_weights",
    "roster",
    "signal_activity",
    "summary",
    "verdict_scorecard",
]

File: quant_rl_trading/dashboard/static/agent_health.js
/* Agent Health — 에이전트가 살아 있는지, 그리고 쓸모가 있는지.
 *
 * 공통 규약은 scope.js 에 있다. 여기서는 이 화면 고유의 표현만 만든다.
 */

async function renderSummary() {
  const body = await fetchJson("agent-health/summary");
  const d = body.data;
  showScope(body);

  document.getElementById("kpis").innerHTML = [
    kpi("Analyst", num(d.total), `측정 완료 ${num(d.measured)}`),
    // M2 완료 기준이 2명이라 이 숫자가 화면의 중심이다.
    kpi("IC 통과", num(d.passed), "M2 기준 2명", d.passed < 2),
    kpi("관찰 모드", num(d.observing), "가중치 0 — 매매 미사용"),
    kpi("가중치 합", dec(d.active_weight, 1), "0이면 아무도 매매에 못 쓴다", d.active_weight === 0),
    kpi("기록된 Signal", num(d.signals), "관찰 모드여도 기록은 남긴다", d.signals === 0),
    kpi("거부 건수", num(d.blocks), "News · SNS"),
  ].join("");

  showAlerts(d.warnings);
}

async function renderRoster() {
  const { data, thresholds } = await fetchJson("agent-health/roster");
  const rows = data.map((item) => {
    const state = !item.measured
      ? `<span class="tag dim">미측정</span>`
      : item.passed
        ? `<span class="tag pass">통과</span>`
        : `<span class="tag observe">관찰</span>`;
    const icCell = item.ic === null
      ? "—"
      : `<span class="${item.passed ? 'good' : 'weak'}">${dec(item.ic)}</span>`;
    return `<tr>
      <td><strong>${item.analyst}</strong><div class="kpi-note">${item.note}</div></td>
      <td>${state}</td>
      <td class="num">${icCell}</td>
      <td class="num">${item.ic_threshold === undefined ? dec(thresholds.ic_threshold ?? null, 2) : dec(item.ic_threshold, 2)}</td>
      <td class="num">${num(item.sample_days)}</td>
      <td class="num">${dec(item.weight, 1)}</td>
      <td class="num">${item.version || "—"}</td>
      <td class="num">${item.measured_at ? item.measured_at.slice(0, 16).replace("T", " ") : "—"}</td>
    </tr>`;
  });

  document.getElementById("roster").innerHTML = `<table>
    <thead><tr>
      <th>Analyst</th><th>상태</th><th class="num">IC</th><th class="num">합격선</th>
      <th class="num">표본(일)</th><th class="num">가중치</th><th class="num">버전</th>
      <th class="num">측정 시각</th>
    </tr></thead>
    <tbody>${rows.join("")}</tbody></table>`;
}

async function renderIcHistory() {
  const { data, thresholds } = await fetchJson("agent-health/ic-history");
  const instance = chart("chart-ic");

  if (!data.series.length) {
    instance.clear();
    instance.setOption({
      ...BASE,
      title: {
        text: "측정 이력 없음 — tools/measure_ic.py --save 로 기록된다",
        left: "center", top: "middle",
        textStyle: { color: COLOR.dim, fontSize: 12, fontWeight: "normal" },
      },
    }, true);
    return;
  }

  const stamps = [...new Set(data.series.flatMap((s) => s.points.map((p) => p.at)))].sort();
  const line = (series, index) => ({
    name: series.analyst,
    type: "line",
    // 점이 하나면 선이 안 보인다. 심볼을 항상 그린다.
    showSymbol: true,
    symbolSize: 7,
    data: stamps.map((at) => {
      const hit = series.points.find((p) => p.at === at);
      return hit ? hit.ic : null;
    }),
    // 계열 구분이지 손익·상태가 아니다 — 공용 범주 팔레트를 쓴다.
    lineStyle: { width: 1.6, color: COLOR.series[index % COLOR.series.length] },
    itemStyle: { color: COLOR.series[index % COLOR.series.length] },
  });

  instance.setOption({
    ...BASE,
    legend: { ...BASE.legend, data: data.series.map((s) => s.analyst) },
    xAxis: { type: "category", data: stamps.map((at) => at.slice(0, 16).replace("T", " ")), ...AXIS },
    yAxis: { type: "value", scale: true, ...AXIS },
    series: [
      ...data.series.map(line),
      {
        // 합격선을 배경에 깐다. 숫자를 눈으로 판정할 수 있어야 한다.
        name: "합격선", type: "line", data: stamps.map(() => thresholds.ic_threshold ?? 0.03),
        showSymbol: false, lineStyle: { width: 1, color: COLOR.warn, type: "dashed" },
      },
    ],
  }, true);
}

async function renderSignals() {
  const { data } = await fetchJson("agent-health/signals");
  const target = document.getElementById("signals");
  if (!data.analysts.length) {
    target.innerHTML = `<div class="empty">기록된 Signal 이 없다.
      관찰 모드여도 점수는 남겨야 나중에 켤 근거가 생긴다.</div>`;
    return;
  }
  target.innerHTML = `<table>
    <thead><tr><th>Analyst</th><th class="num">신호</th><th class="num">종목</th>
      <th class="num">세션</th><th class="num">지연 p90</th><th class="num">최근 as_of</th></tr></thead>
    <tbody>${data.analysts.map((a) => `<tr>
      <td>${a.analyst}<div class="kpi-note">${a.version}</div></td>
      <td class="num">${num(a.signals)}</td>
      <td class="num">${num(a.entities)}</td>
      <td class="num">${num(a.sessions)}</td>
      <td class="num">${ms(a.latency_p90_ms)}</td>
      <td class="num">${a.last_as_of.slice(0, 16).replace("T", " ")}</td>
    </tr>`).join("")}</tbody></table>`;
}

async function renderVerdicts() {
  const { data } = await fetchJson("agent-health/verdicts");
  const target = document.getElementById("verdicts");
  const card = data.scorecard || {};
  if (!data.blocks) {
    target.innerHTML = `<div class="empty">거부 기록이 없다.
      News · SNS 수집기는 M3 실전 직전에 붙인다 — 과거 데이터를 검증할 수 없어
      IC 에 물리지 않는다. 판정 규칙·거부 상한·만료·성적표 배관은 완성돼 있다.</div>`;
    return;
  }
  target.innerHTML = `
    <div class="kpis" style="margin-bottom:10px">
      ${kpi("총 거부", num(data.blocks), "")}
      ${kpi("현재 유효", num(data.active), "expires_at 이 지나면 자동 해제")}
      ${kpi("채점 완료", num(card.settled), `진행 중 ${num(card.pending)}건은 제외`)}
      ${kpi("적중률", pct(card.hit_rate, 1), "차단 종목이 시장보다 더 빠진 비율",
            card.hit_rate !== null && card.hit_rate < 0.5)}
      ${kpi("평균 초과수익", pct(card.mean_excess, 2), "음수면 손실을 피한 것",
            card.mean_excess !== null && card.mean_excess > 0)}
    </div>
    <table>
      <thead><tr><th>사유</th><th class="num">건수</th></tr></thead>
      <tbody>${data.by_category.map((row) => `<tr>
        <td>${row.category}</td><td class="num">${num(row.count)}</td>
      </tr>`).join("")}</tbody>
    </table>`;
}

runAll([renderSummary, renderRoster, renderIcHistory, renderSignals, renderVerdicts]);
